core/src/main/java/org/infinispan/statetransfer/StateTransferManagerImpl.java

```diff
@@ -302,27 +301,29 @@ public boolean isStateTransferInProgress() {
          log.tracef("CommandTopologyId=%s, localTopologyId=%s", cmdTopologyId, localTopologyId);
       }
 
-      if (cmdTopologyId < localTopologyId) {
-         Collection<Address> newTargets = new HashSet<>(cacheTopology.getWriteOwners(affectedKeys));
-         newTargets.remove(rpcManager.getAddress());
-         // Forwarding to the originator would create a cycle
-         // TODO This may not be the "real" originator, but one of the original recipients
-         // or even one of the nodes that one of the original recipients forwarded the command to.
-         // In non-transactional caches, the "real" originator keeps a lock for the duration
-         // of the RPC, so this means we could get a deadlock while forwarding to it.
-         newTargets.remove(origin);
-         if (!newTargets.isEmpty()) {
-            // Update the topology id to prevent cycles
-            command.setTopologyId(localTopologyId);
-            if (log.isTraceEnabled()) {
-               log.tracef("Forwarding command %s to new targets %s", command, newTargets);
-            }
-            // TxCompletionNotificationCommands are the only commands being forwarded now,
-            // and they must be OOB + asynchronous
-            rpcManager.sendToMany(newTargets, command, DeliverOrder.NONE);
+      if (cmdTopologyId >= localTopologyId) {
+         return CompletableFutures.completedEmptyMap();
+      }
+
+      Collection<Address> newTargets = new HashSet<>(cacheTopology.getWriteOwners(affectedKeys));
+      newTargets.remove(rpcManager.getAddress());
+      // Forwarding to the originator would create a cycle
+      // TODO This may not be the "real" originator, but one of the original recipients
+      // or even one of the nodes that one of the original recipients forwarded the command to.
+      // In non-transactional caches, the "real" originator keeps a lock for the duration
+      // of the RPC, so this means we could get a deadlock while forwarding to it.
+      newTargets.remove(origin);
+      if (!newTargets.isEmpty()) {
+         // Update the topology id to prevent cycles
+         command.setTopologyId(localTopologyId);
+         if (log.isTraceEnabled()) {
+            log.tracef("Forwarding command %s to new targets %s", command, newTargets);
          }
+         // TxCompletionNotificationCommands are the only commands being forwarded now,
+         // and they must be OOB + asynchronous
+         rpcManager.sendToMany(newTargets, command, DeliverOrder.NONE);
       }
-      return Collections.emptyMap();
+      return CompletableFutures.completedEmptyMap();
```

Comment:
The `TxCompletionNotificationCommands` is async so, there is nothing to return and this method is no longer used by any other command.
I was in doubt if I should
* return void
* remove the method from `StateTransferManager` since it is only used by `TxCompletionNotificationCommands`

Reply:
Deprecated the method from `StateTransferManager`. :+1: